// @bufbuild/protoplugin/dist/cjs/ecmascript/schema.js
"use strict";
// Copyright 2021-2023 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.toResponse = exports.createSchema = void 0;
const protobuf_1 = require("@bufbuild/protobuf");
const generated_file_js_1 = require("./generated-file.js");
const runtime_imports_js_1 = require("./runtime-imports.js");
const import_symbol_js_1 = require("./import-symbol.js");
const import_path_js_1 = require("./import-path.js");
function createSchema(request, targets, tsNocheck, bootstrapWkt, rewriteImports, importExtension, keepEmptyFiles, pluginName, pluginVersion, pluginParameter) {
    const descriptorSet = (0, protobuf_1.createDescriptorSet)(request.protoFile);
    const filesToGenerate = findFilesToGenerate(descriptorSet, request);
    const runtime = (0, runtime_imports_js_1.createRuntimeImports)(bootstrapWkt);
    const createTypeImport = (desc) => {
        const name = protobuf_1.codegenInfo.localName(desc);
        const from = (0, import_path_js_1.makeImportPath)(desc.file, bootstrapWkt, filesToGenerate);
        return (0, import_symbol_js_1.createImportSymbol)(name, from);
    };
    const generatedFiles = [];
    const schema = {
        targets,
        runtime,
        proto: request,
        files: filesToGenerate,
        allFiles: descriptorSet.files,
        generateFile(name) {
            const genFile = (0, generated_file_js_1.createGeneratedFile)(name, (0, import_path_js_1.deriveImportPath)(name), (importPath) => (0, import_path_js_1.rewriteImportPath)(importPath, rewriteImports, importExtension), createTypeImport, runtime, {
                pluginName,
                pluginVersion,
                pluginParameter,
                tsNocheck,
            }, keepEmptyFiles);
            generatedFiles.push(genFile);
            return genFile;
        },
    };
    return {
        schema,
        getFileInfo() {
            return generatedFiles.flatMap((file) => {
                const fileInfo = file.getFileInfo();
                // undefined is returned if the file has no content
                if (!fileInfo) {
                    return [];
                }
                return [fileInfo];
            });
        },
    };
}
exports.createSchema = createSchema;
function toResponse(files) {
    return new protobuf_1.CodeGeneratorResponse({
        supportedFeatures: protobuf_1.protoInt64.parse(protobuf_1.CodeGeneratorResponse_Feature.PROTO3_OPTIONAL),
        file: files.map((f) => {
            if (f.preamble !== undefined) {
                f.content = f.preamble + "\n" + f.content;
            }
            return f;
        }),
    });
}
exports.toResponse = toResponse;
function findFilesToGenerate(descriptorSet, request) {
    const missing = request.fileToGenerate.filter((fileToGenerate) => descriptorSet.files.every((file) => fileToGenerate !== file.name + ".proto"));
    if (missing.length) {
        throw `files_to_generate missing in the request: ${missing.join(", ")}`;
    }
    return descriptorSet.files.filter((file) => request.fileToGenerate.includes(file.name + ".proto"));
}
